### PR TITLE
Split managed application status and IApplication status

### PR DESCRIPTION
**Pull request by pakrym**
Because of how things were implemented in the past `IAPPLICATION->QueryStatus()` got interleaved with managed application status we use in `INPROCESS_APPLICATION` this is what causes failures like:

```
Failed   Microsoft.AspNetCore.Server.IIS.FunctionalTests.AppOfflineTests.AppOfflineAddedAndRemovedStress(hostingModel: InProcess)
  Error Message:
   Status code was 500
  Expected: True
  Actual:   False
  Stack Trace:
     at Microsoft.AspNetCore.Server.IISIntegration.FunctionalTests.Helpers.<>c__DisplayClass4_0.<<StressLoad>g__RunRequests|0>d.MoveNext() in /_/test/Common.FunctionalTests/Utilities/Helpers.cs:line 45
  --- End of stack trace from previous location where exception was thrown ---
     at Microsoft.AspNetCore.Server.IISIntegration.FunctionalTests.Helpers.StressLoad(HttpClient httpClient, String path, Action`1 action) in /_/test/Common.FunctionalTests/Utilities/Helpers.cs:line 58
     at Microsoft.AspNetCore.Server.IIS.FunctionalTests.AppOfflineTests.AppOfflineAddedAndRemovedStress(HostingModel hostingModel) in /_/test/Common.FunctionalTests/AppOfflineTests.cs:line 194
  --- End of stack trace from previous location where exception was thrown ---

   | [3.592s] Microsoft.AspNetCore.Server.IntegrationTesting.IIS.IISDeployer Information: [aspnetcorev2.dll] Resolving hostfxr parameters for application: C:\Users\appveyor\.dotnet\x64\dotnet.exe arguments: .\InProcessWebSite.dll path: C:\Users\appveyor\AppData\Local\Temp\1\e7ce9b0c468641efb0518f6c2ce60b5f\
   | [3.592s] Microsoft.AspNetCore.Server.IntegrationTesting.IIS.IISDeployer Information: [aspnetcorev2.dll] Process path C:\Users\appveyor\.dotnet\x64\dotnet.exe is dotnet, treating application as portable
   | [3.592s] Microsoft.AspNetCore.Server.IntegrationTesting.IIS.IISDeployer Information: [aspnetcorev2.dll] Resolving absolute path to dotnet.exe from C:\Users\appveyor\.dotnet\x64\dotnet.exe
   | [3.593s] Microsoft.AspNetCore.Server.IntegrationTesting.IIS.IISDeployer Information: [aspnetcorev2.dll] Found dotnet.exe at C:\Users\appveyor\.dotnet\x64\dotnet.exe
   | [3.593s] Microsoft.AspNetCore.Server.IntegrationTesting.IIS.IISDeployer Information: [aspnetcorev2.dll] Resolving absolute path to hostfxr.dll from C:\Users\appveyor\.dotnet\x64\dotnet.exe
   | [3.593s] Microsoft.AspNetCore.Server.IntegrationTesting.IIS.IISDeployer Information: [aspnetcorev2.dll] hostfxr.dll located at C:\Users\appveyor\.dotnet\x64\host\fxr\2.2.0-preview1-26618-02\hostfxr.dll
   | [3.593s] Microsoft.AspNetCore.Server.IntegrationTesting.IIS.IISDeployer Information: [aspnetcorev2.dll] Resolving hostfxr_main arguments application: C:\Users\appveyor\.dotnet\x64\dotnet.exe arguments: .\InProcessWebSite.dll path: C:\Users\appveyor\AppData\Local\Temp\1\e7ce9b0c468641efb0518f6c2ce60b5f\
   | [3.593s] Microsoft.AspNetCore.Server.IntegrationTesting.IIS.IISDeployer Information: [aspnetcorev2.dll] Argument[1] = C:\Users\appveyor\AppData\Local\Temp\1\e7ce9b0c468641efb0518f6c2ce60b5f\InProcessWebSite.dll
   | [3.593s] Microsoft.AspNetCore.Server.IntegrationTesting.IIS.IISDeployer Information: [aspnetcorev2.dll] Loading request handler: C:\Users\appveyor\AppData\Local\Temp\1\e7ce9b0c468641efb0518f6c2ce60b5f\x64\aspnetcorev2_inprocess.dll
   | [3.593s] Microsoft.AspNetCore.Server.IntegrationTesting.IIS.IISDeployer Information: [aspnetcorev2.dll] Creating handler application
   | [3.593s] Microsoft.AspNetCore.Server.IntegrationTesting.IIS.IISDeployer Information: [aspnetcorev2_inprocess.dll] REQUESTHANDLER_CONFIG::GetConfig, set config to ModuleContext
   | [3.593s] Microsoft.AspNetCore.Server.IntegrationTesting.IIS.IISDeployer Information: [aspnetcorev2_inprocess.dll] Resolving hostfxr_main arguments application: C:\Users\appveyor\.dotnet\x64\dotnet.exe arguments: .\InProcessWebSite.dll path: C:\Users\appveyor\AppData\Local\Temp\1\e7ce9b0c468641efb0518f6c2ce60b5f\
   | [3.593s] Microsoft.AspNetCore.Server.IntegrationTesting.IIS.IISDeployer Information: [aspnetcorev2_inprocess.dll] Argument[1] = C:\Users\appveyor\AppData\Local\Temp\1\e7ce9b0c468641efb0518f6c2ce60b5f\InProcessWebSite.dll
   | [3.593s] Microsoft.AspNetCore.Server.IntegrationTesting.IIS.IISDeployer Information: [aspnetcorev2_inprocess.dll] Starting managed application
   | [3.593s] Microsoft.AspNetCore.Server.IntegrationTesting.IIS.IISDeployer Information: [aspnetcorev2_inprocess.dll] Event Log: Application 'C:\Users\appveyor\AppData\Local\Temp\1\e7ce9b0c468641efb0518f6c2ce60b5f\' started the coreclr in-process successfully.
   | [3.593s] Microsoft.AspNetCore.Server.IntegrationTesting.IIS.IISDeployer Information: [aspnetcorev2_inprocess.dll] Starting app_offline monitoring in application C:\Users\appveyor\AppData\Local\Temp\1\e7ce9b0c468641efb0518f6c2ce60b5f\
   | [3.593s] Microsoft.AspNetCore.Server.IntegrationTesting.IIS.IISDeployer Information: [aspnetcorev2_inprocess.dll] Starting file watcher thread
   | [3.593s] Microsoft.AspNetCore.Server.IntegrationTesting.IIS.IISDeployer Information: [aspnetcorev2_inprocess.dll] Received app_offline notification in application C:\Users\appveyor\AppData\Local\Temp\1\e7ce9b0c468641efb0518f6c2ce60b5f\
   | [3.593s] Microsoft.AspNetCore.Server.IntegrationTesting.IIS.IISDeployer Information: [aspnetcorev2_inprocess.dll] Event Log: Application 'C:\Users\appveyor\AppData\Local\Temp\1\e7ce9b0c468641efb0518f6c2ce60b5f\' was recycled after detecting the app_offline file.
   | [3.593s] Microsoft.AspNetCore.Server.IntegrationTesting.IIS.IISDeployer Information: [aspnetcorev2_inprocess.dll] Stopping file watcher thread
   | [3.593s] Microsoft.AspNetCore.Server.IntegrationTesting.IIS.IISDeployer Information: [aspnetcorev2_inprocess.dll] Managed application exited with code 0
   | [3.593s] Microsoft.AspNetCore.Server.IntegrationTesting.IIS.IISDeployer Information: [aspnetcorev2_inprocess.dll] Event Log: Application 'MACHINE/WEBROOT/APPHOST/HTTPTESTSITE20180802151425' has shutdown.
   | [3.593s] Microsoft.AspNetCore.Server.IntegrationTesting.IIS.IISDeployer Information: [aspnetcorev2.dll] Failed HRESULT returned: 0x8007053d at c:\projects\iisintegration\src\aspnetcoremodulev2\aspnetcore\proxymodule.cpp:109 
   | [3.593s] Microsoft.AspNetCore.Server.IntegrationTesting.IIS.IISDeployer Information: [aspnetcorev2.dll] Failed HRESULT returned: 0x8007053d at c:\projects\iisintegration\src\aspnetcoremodulev2\aspnetcore\proxymodule.cpp:125 
   | [3.593s] Microsoft.AspNetCore.Server.IntegrationTesting.IIS.IISDeployer Information: [aspnetcorev2.dll] Failed HRESULT returned: 0x8007053d at c:\projects\iisintegration\src\aspnetcoremodulev2\aspnetcore\proxymodule.cpp:109 
   | [3.593s] Microsoft.AspNetCore.Server.IntegrationTesting.IIS.IISDeployer Information: [aspnetcorev2.dll] Failed HRESULT returned: 0x8007053d at c:\projects\iisintegration\src\aspnetcoremodulev2\aspnetcore\proxymodule.cpp:125 
   | [3.593s] Microsoft.AspNetCore.Server.IntegrationTesting.IIS.IISDeployer Information: [aspnetcorev2.dll] Application went offline
   | [3.593s] Microsoft.AspNetCore.Server.IntegrationTesting.IIS.IISDeployer Information: [aspnetcorev2.dll] Creating handler application
   | [3.593s] Microsoft.AspNetCore.Server.IntegrationTesting.IIS.IISDeployer Information: [aspnetcorev2.dll] Failed HRESULT returned: 0x8007053d at c:\projects\iisintegration\src\aspnetcoremodulev2\aspnetcore\proxymodule.cpp:109 
   | [3.594s] Microsoft.AspNetCore.Server.IntegrationTesting.IIS.IISDeployer Information: [aspnetcorev2.dll] Failed HRESULT returned: 0x8007053d at c:\projects\iisintegration\src\aspnetcoremodulev2\aspnetcore\proxymodule.cpp:125 
   | [3.597s] Microsoft.AspNetCore.Server.IntegrationTesting.IIS.IISDeployer Information: [aspnetcorev2.dll] ASPNET_CORE_GLOBAL_MODULE::OnGlobalStopListening
   | [3.597s] Microsoft.AspNetCore.Server.IntegrationTesting.IIS.IISDeployer Information: [aspnetcorev2.dll] ASPNET_CORE_GLOBAL_MODULE::Terminate
   | [3.631s] Microsoft.AspNetCore.Server.IntegrationTesting.IIS.IISDeployer Information: [Time]: Total time taken for this test variation '3.6301795' seconds
```

APPLICATION_INFO doesn't expect statuses other than RUNNING or RECYCLED:

https://github.com/aspnet/IISIntegration/blob/05eb6a7f7c72bfec63f3c65f63a2f020fe22e705/src/AspNetCoreModuleV2/AspNetCore/proxymodule.cpp#L105-L110

and IN_PROCESS_APPLICATION goes through more states when it shuts down.

So in this PR I've split IAPPLICATION->QueryStatus and managed application status.

### CODE REVIEW

**Comment by pakrym** [comment]
Don't mind tests for now.

File: src/AspNetCoreModuleV2/InProcessRequestHandler/inprocessapplication.h [hunk 1]
```diff
@@ -12,6 +12,15 @@ typedef REQUEST_NOTIFICATION_STATUS(WINAPI * PFN_REQUEST_HANDLER) (IN_PROCESS_HA
 typedef BOOL(WINAPI * PFN_SHUTDOWN_HANDLER) (void* pvShutdownHandlerContext);
 typedef REQUEST_NOTIFICATION_STATUS(WINAPI * PFN_ASYNC_COMPLETION_HANDLER)(void *pvManagedHttpContext, HRESULT hrCompletionStatus, DWORD cbCompletion);
 
+enum MANAGED_APPLICATION_STATUS
```
Comment (jkotalik) [inline src/AspNetCoreModuleV2/InProcessRequestHandler/inprocessapplication.h]: Can this by private?

File: src/AspNetCoreModuleV2/AspNetCore/applicationinfo.cpp [hunk 2]
```diff
@@ -72,14 +72,16 @@ APPLICATION_INFO::GetOrCreateApplication(
...
+            FINISHED(S_OK);
         }
     }
 
     if (AppOfflineApplication::ShouldBeStarted(httpApplication))
     {
         LOG_INFO("Detected app_offline file, creating polling application");
         m_pApplication.reset(new AppOfflineApplication(httpApplication));
+        // Force status update
+        m_pApplication->QueryStatus();
```
Comment (jkotalik) [inline src/AspNetCoreModuleV2/AspNetCore/applicationinfo.cpp]: Why do you need to do this? Isn't the only way we get the status is via QueryStatus? If not, we should redesign this such that we don't need to call QueryStatus.

File: src/AspNetCoreModuleV2/CommonLib/iapplication.h [hunk 3]
```diff
@@ -8,12 +8,8 @@
 
 enum APPLICATION_STATUS
```
Comment (jkotalik) [inline src/AspNetCoreModuleV2/CommonLib/iapplication.h]: Why not just make this a bool like `IsApplicaitonRunning`?

File: src/AspNetCoreModuleV2/CommonLib/application.h [hunk 4]
```diff
@@ -20,24 +19,38 @@ class APPLICATION : public IAPPLICATION
...
+        }
 
         m_fStopCalled = true;
+
+        StopInternal(fServerInitiated);
+    }
+
+    virtual
+    VOID
+    StopInternal(bool fServerInitiated)
```
Comment (jkotalik) [inline src/AspNetCoreModuleV2/CommonLib/application.h]: Feels weird to have an `Internal` function as virtual.

File: src/AspNetCoreModuleV2/RequestHandlerLib/AppOfflineTrackingApplication.cpp [hunk 5]
```diff
@@ -26,11 +24,9 @@ HRESULT AppOfflineTrackingApplication::StartMonitoringAppOffline()
     return hr;
 }
 
-void AppOfflineTrackingApplication::Stop(bool fServerInitiated)
+void AppOfflineTrackingApplication::StopInternal(bool fServerInitiated)
 {
-    APPLICATION::Stop(fServerInitiated);
-
-    m_status = APPLICATION_STATUS::RECYCLED;
+    APPLICATION::StopInternal(fServerInitiated);
```
Comment (jkotalik) [inline src/AspNetCoreModuleV2/RequestHandlerLib/AppOfflineTrackingApplication.cpp]: Should `StopInternal` be called at the end of this function?

Comment (pakrym) [reply src/AspNetCoreModuleV2/RequestHandlerLib/AppOfflineTrackingApplication.cpp]: Why?

File: src/AspNetCoreModuleV2/CommonLib/iapplication.h (hunk 3 above)
Comment (pakrym) [reply src/AspNetCoreModuleV2/CommonLib/iapplication.h]: I was thinking about it but want to leave it as enum for future proofing.

File: src/AspNetCoreModuleV2/CommonLib/application.h (hunk 4 above)
Comment (pakrym) [reply src/AspNetCoreModuleV2/CommonLib/application.h]: DoStop? StopImpl?

Comment (jkotalik) [reply src/AspNetCoreModuleV2/CommonLib/application.h]: Make this one `Stop` and make the call in Application.cpp `StopApplication`?

File: src/AspNetCoreModuleV2/RequestHandlerLib/AppOfflineTrackingApplication.cpp (hunk 5 above)
Comment (jkotalik) [reply src/AspNetCoreModuleV2/RequestHandlerLib/AppOfflineTrackingApplication.cpp]: Nvrm I miss read something the application status. We always want to call the base class before executing our own function. Ex calling `APPLICATION::StopInternal()` before anything we need to do in AppOfflineTrackingApplication.

File: src/AspNetCoreModuleV2/CommonLib/application.h (hunk 4 above)
Comment (pakrym) [reply src/AspNetCoreModuleV2/CommonLib/application.h]: Is it better?

Comment (jkotalik) [reply src/AspNetCoreModuleV2/CommonLib/application.h]: Not enough to actually make that change.

**Pull request merged**